orca_detector/database_parser.py

```diff
@@ -1,95 +1,257 @@
+        # extract audio filenames
+        filenames = [filename for filename in filenames if os.path.splitext(filename)[1].lower() == '.wav']
+        total_files += len(filenames)
+        if len(filenames) == 0:
+            continue  # try next folder
+                
+        # extract folder names as labels from a path that looks like:
+        #   /data/MarineMammalName/1975
+        path, year_folder = os.path.split(dirpath)
+        _, label = os.path.split(path)
+        # strip non alphanumeric characters
+        label = re.sub('\W+','', label )
+            
+        all_samples[label].extend(
+            [os.path.join(dirpath, file) for file in filenames])
+        # print('Loaded data from {}, mapped to label={}'.format(dirpath, label))
 
+    print('Loaded data from {} files. {} labels observed.'.format(total_files, len(all_samples)))
+        
+    train = defaultdict(list)
+    validate = defaultdict(list)
+    test = defaultdict(list)
+
+    # TODO: change train/val/test split to happen *after* quantization
+    
+    # do a stratified train/val/test split
+    for label, files in all_samples.items():
+        if len(files) < 10:
+            continue  # don't both to shuffle
+        random.shuffle(files)
+        num_train_files = int((len(files) + 1) * train_percentage)
+        num_validate_files = int((len(files) + 1) * validate_percentage)
+        train[label] = files[ : num_train_files]
+        validate[label] = files[num_train_files : num_train_files + num_validate_files]
+        test[label] = files[num_train_files + num_validate_files : ]
+
+    # Create lists with each element looking like:
+    #   ['SpermWhale', '/data/SpermWhale/1985/8500901B.wav']
+    train_flattened = [[label, file] 
+                        for label in train.keys() for file in train[label]]
+    validate_flattened = [[label, file] 
+                           for label in train.keys() for file in validate[label]]
+    test_flattened = [[label, file] 
+                       for label in train.keys() for file in test[label]]
+        
+    print('{} sample files in train'.format(len(train_flattened)))
+    print('{} samples files in val'.format(len(validate_flattened)))
+    print('{} samples files in test'.format(len(test_flattened)))
+        
+    train_flattened = _quantize_samples(train_flattened)
+    validate_flattened = _quantize_samples(validate_flattened)
+    test_flattened = _quantize_samples(test_flattened)
+
+    print('\nAfter quantization:')
+    print('{} sample segments in train'.format(len(train_flattened)))
+    print('{} sample segments in val'.format(len(validate_flattened)))
+    print('{} sample segments in test'.format(len(test_flattened)))
+        
+    with open(train_indices_file, 'wb') as fp:
+        pickle.dump(train_flattened, fp)
+        print('Saved training set to {}'.format(train_indices_file))
+
+    with open(validate_indices_file, 'wb') as fp:
```

Comment:
Can I leave this to you as part of performance improvements / pregenerating the numpy arrays?  I felt like I was able to grasp what your code was doing well enough to work with it, but am not comfortable doing any kind of big redesign of it right now.  Resolving this for now, as I think it can be done separately after this PR.

Reply:
Actually I think I misinterpreted your original comment.  I agree about having a method that handles a single file.